### PR TITLE
Symetric, transitive and reflexive equality for intersection types

### LINKED ISSUE #11845: EnsoMultiValue == isn't transitive
As [91987ca](https://github.com/enso-org/enso/pull/11600/commits/91987ca3722ee5320542a20aab7de5d72b7025ba) test shows, `==` on `EnsoMultiValue` isn't transitive.

The behavior of `==` for `EnsoMultiValue` must be redefined.

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Fixes #11845 by comparing all the types an `EnsoMultiValue` _has been cast to_.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, read [it here](https://github.com/enso-org/enso/blob/c17a45f13ce411af4e9bb8ad8c2e2aa861975a3d/docs/types/intersection-types.md#equality--hash-code)
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
- [x] Unit tests have been written where possible.

### CODE REVIEW

**Review by GregoryTravis — APPROVED** [review]

**Comment by JaroslavTulach** [comment]
There is a failure in [test/Base_Tests](https://github.com/enso-org/enso/actions/runs/12376855884/job/34544963171?pr=11897#step:7:3754) line `324` showing that:

https://github.com/enso-org/enso/blob/e4a4bcc61f89620f120349ad42f3724649abc6e2/test/Base_Tests/src/Semantic/Conversion_Spec.enso#L321-L324

`Any.to` method doesn't work as expected for _values with intersection types_ - e.g. the result is not the same as `x:Integer`. Made more consistent in db5353b

File: engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/AnyToTest.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    var conv =
+        ContextUtils.evalModule(ctx, Source.newBuilder("enso", code, "conv.enso").build(), "conv");
+    var both = EnsoMultiValue.create(types, types.length, new Object[] {2L, Text.create("Two")});
+    var eq =
+        ContextUtils.executeInContext(
+            ctx,
+            () -> {
+              var bothValue = ctx.asValue(both);
+              var asIntegerTo = conv.execute(0, bothValue);
+              var asIntegerCast = conv.execute(1, bothValue);
```
Comment (Akirathan) [inline engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/AnyToTest.java:70]: Seems like you forgot to test style 2 and 3, i.e.:
```java
              var asTextTo = conv.execute(2, bothValue);
              var asTextCast = conv.execute(3, bothValue);
assertTrue(equals.execute(asTextTo, asTextCast).asBoolean());
```

**Review by Akirathan — APPROVED** [review]
LGTM. Nice tests. Let's not forget to write benchmarks for intersection types before we start using them in std libs.

File: engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/AnyToTest.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/AnyToTest.java:70]: Good catch, not only it is not called, but it is also broken. These two new tests added in d30da41 are [failing on CI](https://github.com/enso-org/enso/actions/runs/12435358675/job/34721020846?pr=11897#step:7:5435)

- tests ignored in 02080f940d and reported #11935

**Pull request merged**